### ISSUE TITLE
Support Mask RCNN models

### ISSUE THREAD

**Issue by baranorbert** [opencv/opencv#11412]
A model created by Tensorflow/Keras/MS Coco/Mask RCNN drops an exception when I try to import it with OpenCV DNN:

Exception: OpenCV(4.0.0-pre) opencv/modules/dnn/src/tensorflow/tf_importer.cpp:485: error: (-2:Unspecified error) Const kernel input not found in function 'getConstBlob'

##### System information (version)
- OpenCV => 4.0.0-pre
- Operating System / Platform => Ubuntu 16.04 LTS 64 bit
- Compiler => gcc 5.4.0

- Tensorflow => 1.6
- Keras => latest (from pip install)
- Cuda 9.0
- CUDNN: 7.1.2
- MaskRCNN: latest (from git)

##### Detailed description

It seems "fpn_p5upsampled/mul" kernel imp. name is not in the const kernel list (const_kernel_names).
Tensorboard screenshot about the model also attached (it contains the mentioned kernel)

[kernel_names.txt](https://github.com/opencv/opencv/files/1955195/kernel_names.txt)
[const_kernel_names.txt](https://github.com/opencv/opencv/files/1955196/const_kernel_names.txt)
![mask_rcnn](https://user-images.githubusercontent.com/26465525/39364035-c9f750ba-4a2c-11e8-927d-28f4479adbdb.png)

##### Steps to reproduce
1. Train a pre-weighted Coco model (mask_rcnn_coco.h5) => Generate a .h5 format model (mask_rcnn.h5)
2. Export architecture and weights (from mask_rcnn.h5, using Mask_RCNN/scripts/export_model.py) => mask_rcbb.pb (Tensorflow format)
3. Import model with OpenCV dnn
cv::dnn::Net net = cv::dnn::readNetFromTensorflow( "mask_rcnn.pb" );

**Comment by dkurt** [comment]
@baranorbert, Am I right that `mask_rcnn*` is on of the models from https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md#coco-trained-models-coco-models ?

**Comment by baranorbert** [comment]
Hi,

according to the MASK_RCNN readme this is the
mask_rcnn_resnet101_atrous_coco
<http://download.tensorflow.org/models/object_detection/mask_rcnn_resnet101_atrous_coco_2018_01_28.tar.gz>
model.
https://github.com/matterport/Mask_RCNN/blob/master/README.md

Regards,
Norbert



2018-04-27 15:43 GMT+02:00 Dmitry Kurtaev <notifications@github.com>:

**Comment by dkurt** [comment]
@baranorbert, please specify an origin repository. There is no Mask_RCNN/scripts/export_model.py at https://github.com/matterport/Mask_RCNN.

**Comment by baranorbert** [comment]
For the conversion I used this Python script:
https://github.com/ericj974/Mask_RCNN/blob/master/scripts/export_model.py

Parameters:

- Inference config <= config

class InferenceConfig(coco.CocoConfig):
    # Set batch size to 1 since we'll be running inference on
    # one image at a time. Batch size = GPU_COUNT * IMAGES_PER_GPU
    GPU_COUNT = 1
    IMAGES_PER_GPU = 1
    IMAGE_MIN_DIM = 256
    IMAGE_MAX_DIM = 256

config = InferenceConfig()

See: https://github.com/matterport/Mask_RCNN/issues/143

- train_log_dirpath <= [MODEL_DIRECTORY]

- model_filepath  <= [MODEL_DIRECTORY]/mask_rcnn.pb

Is it helps to you if I share my converted model "mask_rcnn.pb"?

If you need any further information feel free to ask me!

Regards,
Norbert




2018-04-27 17:07 GMT+02:00 Dmitry Kurtaev <notifications@github.com>:

**Comment by baranorbert** [comment]
Do you have any progress with this issue? Can I help you somehow?

**Comment by dkurt** [comment]
@baranorbert, we've merged a PR with initial support of Mask-RCNN model from [TensorFlow Object Detection API](https://github.com/tensorflow/models/tree/master/research/object_detection).
